rust/src/ffi/evmc_vm.rs

```diff
@@ -0,0 +1,211 @@
+    };
+
+    let Ok(value) = value.to_str() else {
+        return evmc_set_option_result::EVMC_SET_OPTION_INVALID_VALUE;
+    };
+
+    let mut container = unsafe {
+        // Acquire ownership from EVMC.
+        EvmcContainer::<EvmRs>::from_ffi_pointer(instance)
+    };
+
+    let result = match container.set_option(key, value) {
+        Ok(()) => evmc_set_option_result::EVMC_SET_OPTION_SUCCESS,
+        Err(SetOptionError::InvalidKey) => evmc_set_option_result::EVMC_SET_OPTION_INVALID_NAME,
+        Err(SetOptionError::InvalidValue) => evmc_set_option_result::EVMC_SET_OPTION_INVALID_VALUE,
+    };
+
+    // Release ownership to EVMC.
+    unsafe {
+        EvmcContainer::into_ffi_pointer(container);
+    }
+
+    result
+}
+
+#[no_mangle]
+pub extern "C" fn evmc_create_evmrs() -> *const evmc_vm_t {
+    let new_instance = evmc_vm_t {
+        abi_version: EVMC_ABI_VERSION as i32,
+        destroy: Some(__evmc_destroy),
+        execute: Some(__evmc_execute),
+        get_capabilities: Some(__evmc_get_capabilities),
+        set_option: Some(__evmc_set_option),
+        name: unsafe { CStr::from_bytes_with_nul_unchecked(EVM_RS_NAME.as_bytes()).as_ptr() },
+        version: unsafe { CStr::from_bytes_with_nul_unchecked(EVM_RS_VERSION.as_bytes()).as_ptr() },
+    };
+
+    let container = EvmcContainer::<EvmRs>::new(new_instance);
+
+    // Release ownership to EVMC.
+    unsafe { EvmcContainer::into_ffi_pointer(container) }
+}
+
+extern "C" fn __evmc_destroy(instance: *mut evmc_vm_t) {
+    if instance.is_null() {
+        // This is an irrecoverable error that violates the EVMC spec.
+        process::abort();
+    }
+    unsafe {
+        // Acquire ownership from EVMC. This will deallocate it also at the end of the scope.
+        EvmcContainer::<EvmRs>::from_ffi_pointer(instance);
+    }
+}
+
+extern "C" fn __evmc_execute(
+    instance: *mut evmc_vm_t,
+    host: *const evmc_host_interface,
+    context: *mut evmc_host_context,
+    revision: evmc_revision,
+    message: *const evmc_message,
+    code: *const u8,
+    code_size: usize,
+) -> evmc_result {
+    if instance.is_null()
+        || (host.is_null() && EVMC_CAPABILITY != evmc_capabilities::EVMC_CAPABILITY_PRECOMPILES)
+        || message.is_null()
+        || (code.is_null() && code_size != 0)
+    {
+        // These are irrecoverable errors that violate the EVMC spec.
+        std::process::abort();
+    }
+
+    let execution_message: ExecutionMessage = unsafe {
+        // SAFETY:
+        // message is not null
+        (&*message).into()
+    };
+
+    let code_ref = if code.is_null() {
+        assert_eq!(code_size, 0);
+        &[0u8; 0]
+    } else {
+        // SAFETY:
+        // code is not null and code size > 0
+        unsafe { slice::from_raw_parts(code, code_size) }
+    };
+
+    let container = unsafe {
+        // Acquire ownership from EVMC.
+        EvmcContainer::<EvmRs>::from_ffi_pointer(instance)
+    };
+
+    let result = panic::catch_unwind(|| {
```

Comment:
I have been wondering about a similar mechanism for go as well:
We could spawn a goroutine and check for panics, but currently it is not done. 
I am not sure about the reasons behind, it may have been discarded.

It would be interesting to study what are the implications of doing this, is there any performance penalty?

Reply:
That is certainly true. I don't know the performance implications and would have to read up on this. I kept it like in the macro but I was wondering why not everything in this function in `catch_unwind`. After all there is (unsafe) FFI code outside which if implemented incorrectly would me most likely to panic.

Reply:
You can take a note about this. And you could investigate this later on.